src/lightning/pytorch/core/module.py

```diff
@@ -648,21 +649,28 @@ def forward(self, *args: Any, **kwargs: Any) -> Any:
         """
         return super().forward(*args, **kwargs)
 
-    def training_step(self, *args: Any, **kwargs: Any) -> STEP_OUTPUT:  # type: ignore[return-value]
+    @overload
+    def training_step(self, batch: Any, batch_idx: int, dataloader_idx: int = 0) -> Optional[STEP_OUTPUT]:
+        ...
+
+    @overload
+    def training_step(self, dataloader_iter: Iterable[Dataloader[Any]]) -> Optional[STEP_OUTPUT]:
+        ...
+
+    def training_step(self, *args: Any, **kwargs: Any) -> Optional[STEP_OUTPUT]:
         r"""Here you compute and return the training loss and some additional metrics for e.g. the progress bar or
         logger.
 
         Args:
-            batch (:class:`~torch.Tensor` | (:class:`~torch.Tensor`, ...) | [:class:`~torch.Tensor`, ...]):
```

Comment:
`DataLoader` can output any value, not just `Tensor | tuple[Tensor, ...] | list[Tensor]`. In TorchGeo, we output `dict[str, Tensor]`.